### ISSUE TITLE
'BaseGame with TapDetector' inherits multiple members named 'onTapUp' with incompatible signatures.

### ISSUE THREAD

**Issue by HarrisonQi** [flame-engine/flame#360]
`flame/doc/examples/effects/infinite_effects`

```dart
class MyGame extends BaseGame with TapDetector {
  Square greenSquare;
  Square redSquare;

  MyGame() {
    final green = Paint()..color = const Color(0xAA338833);
    final red = Paint()..color = const Color(0xAA883333);
    greenSquare = Square(green, 100, 100);
    redSquare = Square(red, 200, 200);
    add(greenSquare);
    add(redSquare);
  }

  @override
  void onTapUp(details) {
    greenSquare.clearEffects();
    final dx = details.localPosition.dx;
    final dy = details.localPosition.dy;
    greenSquare.addEffect(MoveEffect(
      destination: Position(dx, dy),
      speed: 250.0,
      curve: Curves.bounceInOut,
      isInfinite: true,
      isAlternating: true,
    ));
    redSquare.clearEffects();
    redSquare.addEffect(ScaleEffect(
      size: Size(dx, dy),
      speed: 250.0,
      curve: Curves.easeInCubic,
      isInfinite: true,
      isAlternating: true,
    ));
  }
}
```

When I run this, console:
```
Launching lib\infinite_effects\main.dart on AOSP on IA Emulator in debug mode...
Running Gradle task 'assembleDebug'...

Compiler message:
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTapUp' with incompatible signatures.
Try adding a declaration of 'onTapUp' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:43:8: Context: This is one of the inherited members.
  void onTapUp(TapUpDetails details) {}
       ^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:55:8: Context: This is one of the inherited members.
  void onTapUp(int pointerId, TapUpDetails details) {
       ^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:43:8: Context: This is one of the inherited members.
  void onTapUp(TapUpDetails details) {}
       ^^^^^^^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTap'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:40:8: Context: The method 'TapDetector.onTap' has fewer positional arguments than those of overridden method 'Game with MultiTouchTapDetector.onTap'.
  void onTap() {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:5:8: Context: This is the overridden method ('onTap').
  void onTap(int pointerId) {}
       ^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTapCancel'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:41:8: Context: The method 'TapDetector.onTapCancel' has fewer positional arguments than those of overridden method 'BaseGame.onTapCancel'.
  void onTapCancel() {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:45:8: Context: This is the overridden method ('onTapCancel').
  void onTapCancel(int pointerId) {
       ^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTapDown'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:42:8: Context: The method 'TapDetector.onTapDown' has fewer positional arguments than those of overridden method 'BaseGame.onTapDown'.
  void onTapDown(TapDownDetails details) {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:50:8: Context: This is the overridden method ('onTapDown').
  void onTapDown(int pointerId, TapDownDetails details) {
       ^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTapUp'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:43:8: Context: The method 'TapDetector.onTapUp' has fewer positional arguments than those of overridden method 'BaseGame.onTapUp'.
  void onTapUp(TapUpDetails details) {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:55:8: Context: This is the overridden method ('onTapUp').
  void onTapUp(int pointerId, TapUpDetails details) {
       ^
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTap' with incompatible signatures.
Try adding a declaration of 'onTap' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:40:8: Context: This is one of the inherited members.
  void onTap() {}
       ^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:4:56: Context: This is one of the inherited members.
import 'package:flame/components/composed_component.dart';
                                                       ^^^^...
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:40:8: Context: This is one of the inherited members.
  void onTap() {}
       ^^^^^
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTapCancel' with incompatible signatures.
Try adding a declaration of 'onTapCancel' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:41:8: Context: This is one of the inherited members.
  void onTapCancel() {}
       ^^^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:45:8: Context: This is one of the inherited members.
  void onTapCancel(int pointerId) {
       ^^^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:41:8: Context: This is one of the inherited members.
  void onTapCancel() {}
       ^^^^^^^^^^^
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTapDown' with incompatible signatures.
Try adding a declaration of 'onTapDown' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:42:8: Context: This is one of the inherited members.
  void onTapDown(TapDownDetails details) {}
       ^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:50:8: Context: This is one of the inherited members.
  void onTapDown(int pointerId, TapDownDetails details) {
       ^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:42:8: Context: This is one of the inherited members.
  void onTapDown(TapDownDetails details) {}
       ^^^^^^^^^

Compiler message:
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTapUp' with incompatible signatures.
Try adding a declaration of 'onTapUp' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:43:8: Context: This is one of the inherited members.
  void onTapUp(TapUpDetails details) {}
       ^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:55:8: Context: This is one of the inherited members.
  void onTapUp(int pointerId, TapUpDetails details) {
       ^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:43:8: Context: This is one of the inherited members.
  void onTapUp(TapUpDetails details) {}
       ^^^^^^^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTap'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:40:8: Context: The method 'TapDetector.onTap' has fewer positional arguments than those of overridden method 'MultiTouchTapDetector.onTap'.
  void onTap() {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:5:8: Context: This is the overridden method ('onTap').
  void onTap(int pointerId) {}
       ^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTapCancel'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:41:8: Context: The method 'TapDetector.onTapCancel' has fewer positional arguments than those of overridden method 'BaseGame.onTapCancel'.
  void onTapCancel() {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:45:8: Context: This is the overridden method ('onTapCancel').
  void onTapCancel(int pointerId) {
       ^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTapDown'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:42:8: Context: The method 'TapDetector.onTapDown' has fewer positional arguments than those of overridden method 'BaseGame.onTapDown'.
  void onTapDown(TapDownDetails details) {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:50:8: Context: This is the overridden method ('onTapDown').
  void onTapDown(int pointerId, TapDownDetails details) {
       ^
lib/infinite_effects/main.dart:17:7: Error: Applying the mixin 'TapDetector' to 'BaseGame' introduces an erroneous override of 'onTapUp'.
class MyGame extends BaseGame with TapDetector {
      ^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:43:8: Context: The method 'TapDetector.onTapUp' has fewer positional arguments than those of overridden method 'BaseGame.onTapUp'.
  void onTapUp(TapUpDetails details) {}
       ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:55:8: Context: This is the overridden method ('onTapUp').
  void onTapUp(int pointerId, TapUpDetails details) {
       ^
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTap' with incompatible signatures.
Try adding a declaration of 'onTap' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:40:8: Context: This is one of the inherited members.
  void onTap() {}
       ^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:5:8: Context: This is one of the inherited members.
  void onTap(int pointerId) {}
       ^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:24:7: Context: This is one of the inherited members.
class BaseGame extends Game with MultiTouchTapDetector {
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:40:8: Context: This is one of the inherited members.
  void onTap() {}
       ^^^^^
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTapCancel' with incompatible signatures.
Try adding a declaration of 'onTapCancel' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:41:8: Context: This is one of the inherited members.
  void onTapCancel() {}
       ^^^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:45:8: Context: This is one of the inherited members.
  void onTapCancel(int pointerId) {
       ^^^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:41:8: Context: This is one of the inherited members.
  void onTapCancel() {}
       ^^^^^^^^^^^
lib/infinite_effects/main.dart:17:7: Error: Class 'BaseGame with TapDetector' inherits multiple members named 'onTapDown' with incompatible signatures.
Try adding a declaration of 'onTapDown' to 'BaseGame with TapDetector'.
class MyGame extends BaseGame with TapDetector {
      ^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:42:8: Context: This is one of the inherited members.
  void onTapDown(TapDownDetails details) {}
       ^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/game/base_game.dart:50:8: Context: This is one of the inherited members.
  void onTapDown(int pointerId, TapDownDetails details) {
       ^^^^^^^^^
/D:/flutter/.pub-cache/hosted/pub.flutter-io.cn/flame-0.21.0/lib/gestures.dart:42:8: Context: This is one of the inherited members.
  void onTapDown(TapDownDetails details) {}
       ^^^^^^^^^
Target kernel_snapshot failed: Exception: Errors during snapshot creation: null
build failed.

FAILURE: Build failed with an exception.

* Where:
Script 'D:\flutter\packages\flutter_tools\gradle\flutter.gradle' line: 896

* What went wrong:
Execution failed for task ':app:compileFlutterBuildDebug'.
> Process 'command 'D:\flutter\bin\flutter.bat'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 10s
Exception: Gradle task assembleDebug failed with exit code 1
```
flutter doctor:
```
C:\Users\10361>flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel master, 1.19.0-2.0.pre.143, on Microsoft Windows [Version 10.0.18362.836], locale zh-CN)

[√] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[√] Chrome - develop for the web
[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.6.0)
[√] Android Studio (version 3.5)
[√] IntelliJ IDEA Ultimate Edition (version 2019.3)
[√] Connected device (4 available)

• No issues found!
```

Is my code or environment incompatible?

I also tried to find how to "adding a declaration of 'onTapUp' to 'BaseGame with TapDetector'", but I did not find a solution...

**Comment by erickzanardo** [comment]
You can’t use TapDetector on BaseGame on 0.21. This is already fixed on the develop branch and will be released on 0.22 soon

**Comment by HarrisonQi** [comment]
After upgrading to` 0.22.0`, it runs successfully👍👍👍
